README.md
# Order Statistic Tree

This repository contains an implementation of order statistic tree in Haskell programming language.

I could not find an order statistic tree at Hackage, so I have to develop one.

This implementation uses weight-balanced trees as desribed in
- Hirai, Yoichi, and Kazuhiko Yamamoto. "Balancing weight-balanced trees." Journal of Functional Programming 21.03 (2011): 287-307.

Also some of its code is based on code from containers package.

Implementation of order statistic tree is described in
- Cormen, T.H., Leiserson, Rivest, Stein. Introduction to algorithms. The MIT Press. 3rd ed.

# Installation

This package will be deployed to hackage, so you can install it using cabal:

```
cabal install order-statistic-tree
```

# Building

```
cabal configure
cabal build
```

# Testing

```
cabal configure --enable-tests --enable-benchmarks
cabal test
```

# Benchmarks

I tried to make this tree as fast as possible. I'm not bos, but results on my i7-4790 with 16Gb RAM are following:
- OSTree was created from 1.000.000 random numbers in 2.087 ± 0.021 s (e.g. for Data.Map.Strict - 1.977 ± 0.016 s);
- deletion from OSTree with 1.000.000 random numbers was made in 13.94 ± 0.93 ms;
- lookup from OSTree with 1.000.000 random numbers was made in 208.2 ± 3.48 ns;
- selection from OSTree with 1.000.000 random numbers was made in 92.72 ± 1.91 ns;
- full testing protocol can be found in result-bench.txt.

```
cabal configure --enable-tests --enable-benchmarks
cabal bench
```

If someone knows how to improve these results or benchmarking itself, please don't hesitate to contact me



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Package Versioning Policy](https://pvp.haskell.org).

## [Unreleased]
### Added
### Changed
### Deprecated
### Removed
### Fixed
### Security

## 0.1.1.1 - 2018-11-02

### Other
- updated dependency range
- add stack.yaml

## 0.1.1.0 - 2016-03-02

### Added
- deleteFindMin and deleteFindMax operations
- asymptotics of running time
- benchmark results, readme etc.

## 0.1.0.1 - 2016-03-01

Added documentation

## 0.1.0.0 - 2016-02-29

Initial release





test/Spec.hs

import Test.Tasty
import Test.Tasty.HUnit
import Test.Tasty.QuickCheck

import Data.List
import Data.Maybe
import qualified Data.OSTree as M
import Data.OSTree.Types
import qualified Data.OSTree.Internal as MI

main = defaultMain tests

-- instance Arbitrary OSTree where
--   arbitrary = sized $ \n -> do
--     k <- 
  
--   shrink Tip = []
--   shring (Bin _ k l r) =
--     [Tip] ++
--     [l, r] ++
--     [bin k' l' r' | (k',l',r') <- shrink (k,l,r)]

prepareInsert :: [Int] -> [Int] -> (OSTree Int, [Int])
prepareInsert inserted notInserted = let
  toInsert = nub inserted \\ nub notInserted :: [Int]
  m = foldr M.insert M.empty toInsert
  in (m, toInsert)

prepareDelete :: [Int] -> [Int] -> OSTree Int
prepareDelete inserted deleted = let
  toInsert = nub $ inserted ++ deleted :: [Int]
  m' = foldr M.insert M.empty toInsert
  in foldr M.delete m' deleted

(!!?) :: [a] -> Int -> Maybe a
xs !!? k
  | 0 <= k && k < length xs = Just $ xs !! k
  | otherwise               = Nothing

tests :: TestTree
tests = testGroup "Tests"
        [ testGroup "Insertion"
          [ testProperty "insert inserts all elements" $ \inserted notInserted -> let (m, toInsert) = prepareInsert inserted notInserted
                                                                                  in all (isJust . flip M.lookup m) toInsert &&
                                                                                     all (isNothing . flip M.lookup m) notInserted
          , testProperty "insert leave tree balanced" $ \inserted notInserted -> let (m, toInsert) = prepareInsert inserted notInserted
                                                                                 in MI.balanced m
          , testProperty "insert leave tree correctly sized" $ \inserted notInserted -> let (m, toInsert) = prepareInsert inserted notInserted
                                                                                        in MI.count m === M.size m
          ]
        , testGroup "Deletion"
          [ testProperty "delete deletes" $ \inserted deleted -> let m = prepareDelete inserted deleted
                                                                 in all (isNothing . flip M.lookup m) deleted
          , testProperty "delete leave tree correctly sized" $ \inserted deleted -> let m = prepareDelete inserted deleted
                                                                                    in MI.count m === M.size m 
          ]
        , testGroup "Conversion to List"
          [ testProperty "toList correctly converts" $ \inserted -> let m = foldr M.insert M.empty (inserted :: [Int])
                                                                    in M.toList m === sort inserted
          , testCase "toList (Bin 1 Tip Tip)" $ M.toList (Bin 1 1 Tip Tip) @?= [1]
          ]
        , testGroup "Conversion from List"
          [ testProperty "toList . fromList == id" $ \list -> M.toList (M.fromList list) === sort (list :: [Int])
          , testCase "fromList [8,8]" $ M.fromList [8,8] @?= Bin 2 8 Tip (Bin 1 8 Tip Tip)
          ]
        , testGroup "Selection"
          [ testProperty "select selects correct item" $ \list k -> M.select (M.fromList (list :: [Int])) (k::Int) === sort list !!? (k-1)
          ] 
        -- , testGroup "Ranking"
        --   [ testProperty "rank ranks correctly when item is presented in tree" $ \list k -> (0<=k && k<length list) ==> let
        --        sorted = sort $ nub list :: [Int]
        --        item = sorted !! k
        --        in M.rank item (M.fromList list) === Just k
        --   , testProperty "rank ranks correctly when item is not presented in tree" $ \list item -> item`notElem`list ==> let
        --        sorted = sort $ nub list :: [Int]
        --        in M.rank item (M.fromList list) === Nothing
        --   ]
        ]



bench/Bench.hs
{-# LANGUAGE DeriveGeneric, DeriveAnyClass #-}

import Criterion.Main
import Control.Arrow
import Control.DeepSeq (NFData)
import Control.Monad
import System.Random

import qualified Data.Set as Set
import qualified Data.OSTree as M
import Data.OSTree (OSTree)

instance NFData a => NFData (OSTree a)

type Env = [Int]

sortedItems :: Int -> IO Env
sortedItems n = do
  return [1000..1000+n]

revSortedItems :: Int -> IO Env
revSortedItems = (reverse <$>) . sortedItems

randomItems :: Int -> IO Env
randomItems n = replicateM n randomIO

tree :: IO Env -> IO (OSTree Int, Int)
tree items = do
  tree <- M.fromList <$> items
  item <- (`mod` M.size tree) <$> randomIO
  return (tree,item)

allKs = [10000,20000,50000,100000,200000,1000000]

envs :: [(String, Int -> IO Env)]
envs = [ ("Sorted", sortedItems), ("Reverse", revSortedItems), ("Random", randomItems) ]

envs' :: [(String, Int -> IO (OSTree Int, Int))]
envs' = map (second (tree.)) envs

onAllEnv :: (NFData a) =>
            [(String, Int -> IO a)] ->
            [Int] ->
            (String -> a -> Benchmark) ->
            [Benchmark]
onAllEnv envs ks test = [ bgroup (show k)
                          [ env (getEnv k) $ \ ~environ ->
                             test name environ
                          | (name, getEnv) <- envs]
                        | k <- ks ]

main = defaultMain
       [ bgroup "Insertions" $ onAllEnv envs  allKs insertions
       , bgroup "Deletions"  $ onAllEnv envs' allKs deletions
       , bgroup "Lookups"    $ onAllEnv envs' allKs lookups
       , bgroup "Selections" $ onAllEnv envs' allKs selections
       , bgroup "Data.Map.Strict.insert" $ onAllEnv envs [last allKs] strictMapInsert
       ]


insertions :: String -> Env -> Benchmark
insertions name ~environ = bench name $ nf M.fromList environ

deletions :: String -> (OSTree Int, Int) -> Benchmark
deletions name ~(tree,item) = bench name $ nf (M.delete item) tree

lookups :: String -> (OSTree Int, Int) -> Benchmark
lookups name ~(tree,item) = bench name $ nf (M.lookup item) tree

selections :: String -> (OSTree Int, Int) -> Benchmark
selections name ~(tree,item) = bench name $ nf (flip M.select item) tree

strictMapInsert :: String -> Env -> Benchmark
strictMapInsert name ~environ = bench name $ nf Set.fromList environ



src/Data/OSTree/Internal.hs
module Data.OSTree.Internal where

import Data.OSTree.Types

-- | Returns size of the tree
size :: OSTree a -> Size
size Tip = 0
size (Bin sz _ _ _) = sz

bin :: a -> OSTree a -> OSTree a -> OSTree a
bin k l r = Bin (size l + size r + 1) k l r

balanced :: OSTree a -> Bool
balanced Tip = True
balanced (Bin _ _ l r) = isBalanced l r && isBalanced r l
                      && balanced l     && balanced r

count :: OSTree a -> Size
count Tip = 0
count (Bin _ _ l r) = 1 + count l + count r


-- balances

balanceL :: a -> OSTree a -> OSTree a -> OSTree a
balanceL k l r
  | isBalanced l r = bin k l r
  | otherwise      = rotateL k l r

balanceR :: a -> OSTree a -> OSTree a -> OSTree a
balanceR k l r
  | isBalanced r l = bin k l r
  | otherwise      = rotateR k l r

-- rotation L

rotateL :: a -> OSTree a -> OSTree a -> OSTree a
rotateL k l r@(Bin _ _ rl rr)
  | isSingle rl rr = singleL k l r
  | otherwise      = doubleL k l r
rotateL _ _ _      = error "rotateL"

singleL :: a -> OSTree a -> OSTree a -> OSTree a
singleL k1 t1 (Bin _ k2 t2 t3) = bin k2 (bin k1 t1 t2) t3
singleL _ _ _                  = error "singleL"

doubleL :: a -> OSTree a -> OSTree a -> OSTree a
doubleL k1 t1 (Bin _ k2 (Bin _ k3 t2 t3) t4) =
  bin k3 (bin k1 t1 t2) (bin k2 t3 t4)

-- rotation R

rotateR :: a -> OSTree a -> OSTree a -> OSTree a
rotateR k l@(Bin _ _ ll lr) r
  | isSingle lr ll = singleR k l r
  | otherwise      = doubleR k l r
rotateR _ _ _      = error "rotateR"

singleR :: a -> OSTree a -> OSTree a -> OSTree a
singleR k1 (Bin _ k2 t1 t2) t3 = bin k2 t1 (bin k1 t2 t3)
singleR _ _ _                  = error "singleR"

doubleR :: a -> OSTree a -> OSTree a -> OSTree a
doubleR k1 (Bin _ k2 t1 (Bin _ k3 t2 t3)) t4 =
  bin k3 (bin k2 t1 t2) (bin k1 t3 t4)

-- config

isBalanced :: OSTree a -> OSTree a -> Bool
isBalanced a b = 3 * (size a + 1) >= size b + 1

isSingle :: OSTree a -> OSTree a -> Bool
isSingle a b = size a + 1 < 2 * (size b + 1)


-- | /O(log n)/. Delete and find the minimal element.
--
-- > deleteFindMin (fromList [(5,"a"), (3,"b"), (10,"c")]) == ((3,"b"), fromList[(5,"a"), (10,"c")])
-- > deleteFindMin                                            Error: can not return the minimal element of an empty map

deleteFindMin :: OSTree a -> (a,OSTree a)
deleteFindMin t
  = case t of
      Bin _ k Tip r -> (k,r)
      Bin _ k l r   -> let (km,l') = deleteFindMin l in (km,balance k l' r)
      Tip           -> (error "Map.deleteFindMin: can not return the minimal element of an empty map", Tip)

-- | /O(log n)/. Delete and find the maximal element.
--
-- > deleteFindMax (fromList [(5,"a"), (3,"b"), (10,"c")]) == ((10,"c"), fromList [(3,"b"), (5,"a")])
-- > deleteFindMax empty                                      Error: can not return the maximal element of an empty map

deleteFindMax :: OSTree a -> (a,OSTree a)
deleteFindMax t
  = case t of
      Bin _ k l Tip -> (k,l)
      Bin _ k l r   -> let (km,r') = deleteFindMax r in (km,balance k l r')
      Tip           -> (error "Map.deleteFindMax: can not return the maximal element of an empty map", Tip)

glue :: OSTree a -> OSTree a -> OSTree a
glue Tip r = r
glue l Tip = l
glue l r
  | size l > size r = let (km,l') = deleteFindMax l in balanceL km l' r
  | otherwise       = let (km,r') = deleteFindMin r in balanceR km l r'

balance :: a -> OSTree a -> OSTree a -> OSTree a
balance k l r
  | isBalanced l r && isBalanced r l = bin k l r
  | size l > size r                  = rotateR k l r
  | otherwise                        = rotateL k l r



src/Data/OSTree.hs
{-|
Module      : Data.OSTree
Description : Order Statistic Tree
Copyright   : (c) Lambda Kazan, 2016
License     : BSD3
Maintainer  : mz@lambdasoft.ru
Stability   : experimental
Portability : POSIX

= Order Statistic Tree

This implementation uses weight-balanced trees which are desribed in

- Hirai, Yoichi, and Kazuhiko Yamamoto. "Balancing weight-balanced trees." Journal of Functional Programming 21.03 (2011): 287-307.

Also some of its code is based on containers package.

Implementation of order statistic tree is described in

- Cormen, T.H., Leiserson, Rivest, Stein. Introduction to algorithms. The MIT Press. 3rd ed.

= Benchmarks

I tried to make this tree as fast as possible. I'm not bos, but results on my i7-4790 with 16Gb RAM are following:

 * OSTree was created from 1.000.000 random numbers in 2.087 ± 0.021 s (e.g. for Data.Map.Strict - 1.977 ± 0.016 s);
 * deletion from OSTree with 1.000.000 random numbers was made in 13.94 ± 0.93 ms;
 * lookup from OSTree with 1.000.000 random numbers was made in 208.2 ± 3.48 ns;
 * selection from OSTree with 1.000.000 random numbers was made in 92.72 ± 1.91 ns;
 * full testing protocol can be found in result-bench.txt.

@
cabal configure --enable-tests --enable-benchmarks
cabal bench
@

If someone knows how to improve these results or benchmarking itself, please don't hesitate to contact me

-}
module Data.OSTree
       ( OSTree
         -- * Creating OSTree
       , empty
       , singleton
         -- * Search Tree operations
       , size
       , insert
       , lookup
       , delete
       , deleteFindMin
       , deleteFindMax
         -- * Conversions
       , toList
       , fromList
         -- * Statistics
       , select
       )
       where

import Prelude hiding (lookup)
import Data.List (foldl')

import Data.OSTree.Types
import Data.OSTree.Internal

-- https://yoichihirai.com/bst.pdf

-- | /O(1)/. Returns an empty tree
empty :: OSTree a
empty = Tip

-- | /O(1)/. Returns a tree with single element
singleton :: a -> OSTree a
singleton k = Bin 1 k Tip Tip

-- | /O(log n)/. Insert the element into the tree
insert :: (Ord a) => a -> OSTree a -> OSTree a
insert kx Tip = singleton kx
insert kx t@(Bin sz ky l r) = case compare kx ky of
  LT -> balanceR ky (insert kx l) r
  GT -> balanceL ky l (insert kx r)
  EQ -> balanceR ky l (insert kx r)

-- | /O(log n)/. Lookup the element in the tree
lookup :: (Ord a) => a -> OSTree a -> Maybe a
lookup kx Tip = Nothing
lookup kx (Bin _ ky l r) = case compare kx ky of
  LT -> lookup kx l
  GT -> lookup kx r
  EQ -> Just ky

-- | /O(log n)/. Delete first occurence of the element from the tree
delete :: (Ord a) => a -> OSTree a -> OSTree a
delete k t
  = case t of
      Tip -> Tip
      Bin _ kx l r
          -> case compare k kx of
               LT -> balanceL kx (delete k l) r
               GT -> balanceR kx l (delete k r)
               EQ -> glue l r

-- | /O(n)/. Return list of elements of the tree
toList :: OSTree a -> [a]
toList k = toListL k []

toListL :: OSTree a -> [a] -> [a]
toListL Tip = id
toListL (Bin _ k l r) = toListL l . (k:) . toListL r

-- | /O(n log n)/. Convert list of elements to the tree
fromList :: (Ord a) => [a] -> OSTree a
fromList = foldl' (flip insert) empty

-- | /O(log n)/. Returns i-th least element of the tree
select :: OSTree a              -- ^ tree
       -> Int                   -- ^ index 'i', starting from 1
       -> Maybe a               -- ^ if there are at least 'i' elements, returns i-th least element
select Tip i = Nothing
select (Bin _ k l r) i = let
  n = size l + 1
  in case compare i n of
    EQ -> Just k
    LT -> select l i
    GT -> select r $ i-n


